### PR TITLE
Report exceptions we encounter in the package search update service.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest vsi please

File: src/VisualStudio/Core/Def/Packaging/PackageSearchService.Update.cs [hunk 1]
```diff
@@ -560,10 +560,17 @@ private AddReferenceDatabase CreateDatabaseFromBytes(byte[] bytes)
...
                         action();
-                        return;
                     }
-                    catch (Exception e) when (_service._swallowException(e))
+                    catch (Exception e)
                     {
+                        // Normal IO exception. Don't bother reporting it.  We don't want to get
+                        // lots of hits just because we couldn't write a file because of something
+                        // like an anti-virus tool lockign the file.
+                        if (!IOUtilities.IsNormalIOException(e))
```
Comment (balajikris) [inline src/VisualStudio/Core/Def/Packaging/PackageSearchService.Update.cs]: typo at 'lockign' but dont bother.

**Comment by balajikris** [comment]
:+1:

File: src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs [hunk 2]
```diff
@@ -23,26 +23,21 @@ public static T PerformIO<T>(Func<T> function, T defaultValue = default(T))
...
 
             return defaultValue;
         }
+
+        public static bool IsNormalIOException(Exception e)
+        {
+            return e is IOException ||
+                   e is SecurityException ||
+                   e is ArgumentException ||
+                   e is UnauthorizedAccessException ||
```
Comment (Pilchie) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: Don't we already have a helper for this somewhere in Workspaces?

The compiler build task already has `IsIoRelatedException`, which does `(e is ArgumentException && !(e is ArgumentNullException))`.

Should this do the same.

**Comment by Pilchie** [comment]
:+1:

**Comment by jmarolf** [comment]
:+1:

File: src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs [hunk 3]
```diff
@@ -23,26 +23,21 @@ public static T PerformIO<T>(Func<T> function, T defaultValue = default(T))
...
             }
 
             return defaultValue;
         }
+
+        public static bool IsNormalIOException(Exception e)
+        {
+            return e is IOException ||
+                   e is SecurityException ||
+                   e is ArgumentException ||
```
Comment (jaredpar) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: I wouldn't agree `ArgumentException` is a normal IO exception, it represents a programming error.

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: i'm not sure htat's true when it comes to the windows FS apis.  for example, you might try to name something "con" and you could get an ArgumentException.  There is no APi that says "this name is ok", so you always run the risk of having a name that will cause an exception.

File: src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: This is hte workspace helper :)

I'm pretty sure our version has lived longer.  Perhaps the compiler build task should switch to ours?

File: src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs (hunk 3 above)
Comment (Pilchie) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: Like it or not, all of our code that uses IO seems to handle `ArgumentException`, so for safety I would be consistent here.

Comment (jaredpar) [inline src/Features/Core/Portable/Shared/Utilities/IOUtilities.cs]: Sorry I was reading this backwards. I was reading this as an exclusion list. 

IMHO: just catch exception.

**Comment by dpoeschl** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz

**Comment by MattGertz** [comment]
Approved pending successful completion of tests.

**Pull request merged**